[FIX] Scheduler crashloopbackoff when using `hostname_callable`

V0lantis opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
Fix a scheduler `crashLoopBackOff` when `hostname_callable` was defined with a function which returned something different than the [`hostname`](https://linux.die.net/man/1/hostname) command from linux.

## Context
In our production Airflow, I updated the [`hostname_callable`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#hostname-callable) config with [`airflow.utils.net.get_host_ip_address`](https://github.com/apache/airflow/blob/a3f2df0f45973ddb97990361fdc5caa256c175ca/airflow/utils/net.py#L46-L48) so that we could read the pod logs produced by our celery workers. But this new config was making kubernetes to restart the scheduler pods because the [command defined in helm](https://github.com/apache/airflow/blob/451181f6f2efd8b5e6bd7807e84a17d48a4fe875/chart/templates/_helpers.yaml#L640-L641) was returning a "No alive jobs found."
I therefore updated the cli command jobs to allow the use of `--use-hostname-callable`  in:
```bash
airflow jobs check --job-type SchedulerJob --use-hostname-callable
```
to use the `hostname_callable` that the user defined in its Airflow conf. Also updated the helm chart to with this new conf since I think it will be more reliable than to use linux `hostname` command.

jedcunningham commented on airflow/cli/commands/jobs_command.py:
```diff
@@ -36,6 +40,9 @@ def check(args, session=None):
         query = query.filter(BaseJob.hostname == args.hostname)
+    if args.use_hostname_callable:
+        hostname = get_hostname()
+        query = query.filter(BaseJob.hostname == hostname)
```
```suggestion
        query = query.filter(BaseJob.hostname == get_hostname())
```

nit

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -638,7 +638,7 @@ Create the name of the cleanup service account to use
   - |
     CONNECTION_CHECK_MAX_COUNT=0 AIRFLOW__LOGGING__LOGGING_LEVEL=ERROR exec /entrypoint \
-    airflow jobs check --job-type SchedulerJob --hostname $(hostname)
+    airflow jobs check --job-type SchedulerJob
```
```suggestion
    airflow jobs check --job-type SchedulerJob --use-hostname-callback
```

This also needs to be done conditionally based on `airflowVersion`.

V0lantis replied:
True thank-you :+1:

jedcunningham replied:
Don't forget, this won't be released until 2.4.0, so you'll need to make using that over `--hostname` conditional.

jedcunningham commented on airflow/cli/cli_parser.py:
```diff
@@ -897,6 +897,13 @@ def string_lower_type(val):
 )
 
+ARG_JOB_HOSTNAME_CALLABLE_FILTER = Arg(
+    ("--use-hostname-callable",),
```
Not sure I love `use-hostname-callable`. Maybe `local-only`? @dstandish @ephraimbuddy

V0lantis replied:
It's a nice suggestion @jedcunningham. I don't like so much `use-hostname-callable` either, but I am not sure `local-only` would be very _telling_.

ephraimbuddy replied:
What about not adding a new arg but having the hostname arg default to `get_hostname`?
i.e
```python
ARG_JOB_HOSTNAME_FILTER = Arg(
    ("--hostname",),
    default=get_hostname(),
    type=str,
    help="The hostname of job(s) that will be checked.",
)
```
The effect of this would be that the CLI check command would always check with hostname instead of none.

As per the name, what do you think about `--use-configured-hostname` or `--use-default-hostname`

V0lantis replied:
Sounds very good, thanks for the proposal. Going to implement this

ephraimbuddy commented on chart/templates/_helpers.yaml:
```diff
@@ -638,7 +638,7 @@ Create the name of the cleanup service account to use
   - |
     CONNECTION_CHECK_MAX_COUNT=0 AIRFLOW__LOGGING__LOGGING_LEVEL=ERROR exec /entrypoint \
-    airflow jobs check --job-type SchedulerJob --hostname $(hostname)
+    airflow jobs check --job-type SchedulerJob --use-hostname-callback
```
```suggestion
    airflow jobs check --job-type SchedulerJob --use-hostname-callable
```

V0lantis replied:
thanks :+1:

ephraimbuddy commented on airflow/cli/commands/jobs_command.py:
```diff
@@ -27,6 +27,7 @@ def check(args, session=None):
     """Checks if job(s) are still alive"""
     if args.allow_multiple and not args.limit > 1:
         raise SystemExit("To use option --allow-multiple, you must set the limit to a value greater than 1.")
+
```
If we want to use `default` we need to remove this condition, it's not useful again, right?
```python
if args.hostname:
        query = query.filter(BaseJob.hostname == args.hostname)
```
and add a filter by hostname on the query below

V0lantis replied:
The `if args.hostname:` is indeed not necessary anymore. but the filter is working fine as it is now, or I am not sure to understand your point :)

ephraimbuddy replied:
Just a nit, we could have this:
```python
 query = (
        session.query(BaseJob)
        .filter(BaseJob.state == State.RUNNING)
        .filter(BaseJob.hostname == args.hostname)
        .order_by(BaseJob.latest_heartbeat.desc())
    )
```
WDYT?

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -638,7 +638,7 @@ Create the name of the cleanup service account to use
   - |
     CONNECTION_CHECK_MAX_COUNT=0 AIRFLOW__LOGGING__LOGGING_LEVEL=ERROR exec /entrypoint \
-    airflow jobs check --job-type SchedulerJob --hostname $(hostname)
+    airflow jobs check --job-type SchedulerJob --local
```
This needs to be conditional on 2.4.0 or later.

V0lantis replied:
Yep, I did it in a previous commit and removed it. thanks. Re added it in dc45752cb037fb68e06de3e8e49e7ad59caefa00

V0lantis committed:
[FEAT] Fix a scheduler crashLoopBackOff

Fix a scheduler crashLoopBackOff when `hostname_callable` was defined
with a function which returned something different than the hostname
command from linux.

jedcunningham commented on airflow/cli/cli_parser.py:
```diff
@@ -905,6 +905,13 @@ def string_lower_type(val):
+    ("--local",),
+    action='store_true',
+    help="If passed, this command will use the `hostname_callable` function"
+    "defined in Airflow configuration.",
```
```suggestion
    help="If passed, this command will only show jobs from the local host "
    "(those with a hostname matching what `hostname_callable` returns).",
```

I think this reads a little better?

jedcunningham commented on airflow/cli/cli_parser.py:
```diff
@@ -1779,13 +1786,23 @@ class GroupCommand(NamedTuple):
             'examples:\n'
             'To check if the local scheduler is still working properly, run:\n'
             '\n'
             '    $ airflow jobs check --job-type SchedulerJob --hostname "$(hostname)"\n'
```
```suggestion
            '    $ airflow jobs check --job-type SchedulerJob --local"\n'
```

jedcunningham commented on airflow/cli/cli_parser.py:
```diff
@@ -1779,13 +1786,23 @@ class GroupCommand(NamedTuple):
+            'To check if the local scheduler is still working properly with `hostname_callback`, run:\n'
+            '\n'
+            '    $ airflow jobs check --job-type SchedulerJob --local\n'
+            '\n'
```
```suggestion
```

I think the single example is enough.

However, if we wanted to, we could add a hostname example for a different host.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -633,7 +633,13 @@ Create the name of the cleanup service account to use
 {{define "scheduler_liveness_check_command"}}
 
-  {{- if semverCompare ">=2.1.0" .Values.airflowVersion }}
+  {{- if semverCompare ">=2.4.0" .Values.airflowVersion }}
```
```suggestion
  {{- if semverCompare ">=2.5.0" .Values.airflowVersion }}
```

Sorry, we missed 2.4.0, and this will ship in 2.5.0 instead.

jedcunningham commented on chart/templates/_helpers.yaml:
```diff
@@ -659,11 +665,19 @@ Create the name of the cleanup service account to use
-  - |
-    CONNECTION_CHECK_MAX_COUNT=0 AIRFLOW__LOGGING__LOGGING_LEVEL=ERROR exec /entrypoint \
-    airflow jobs check --job-type TriggererJob --hostname $(hostname)
+    {{- if semverCompare ">=2.4.0" .Values.airflowVersion }}
```
```suggestion
    {{- if semverCompare ">=2.5.0" .Values.airflowVersion }}
```

V0lantis committed:
Apply suggestions from code review

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
